packages/cardano-services/src/WsServer/server.ts

```diff
@@ -0,0 +1,278 @@
+
+        if (!payload) throw new Error('Missing payload in NOTIFY');
+
+        // The payload of the NOTIFY event contain the tip in the correct format
+        this.tip = JSON.parse(payload);
+      } catch (error) {
+        return this.logger.error(error, 'Error while handling tip notification');
+      }
+
+      const message: WSMessage = { tip: this.tip };
+
+      // Create only once a stringified version of a message with the new tip
+      const stringMessage = JSON.stringify(message);
+      // Propagate the new tip to all the connected clients
+      for (const ws of this.wss.clients.values()) ws.send(stringMessage);
+    };
+  }
+
+  private listenNotify() {
+    const { db, logger } = this;
+
+    // This is the function which listens for events throw the NOTIFY command.
+    // It recursively calls itself to handle reconnection.
+    const addListener = () => {
+      logger.info('Connecting to DB to listen on sdk_tip NOTIFY...');
+
+      // Ask for a DB client to the connections pool.
+      db.connect((err, client, done) => {
+        if (err) {
+          logger.error(err, 'Error while connecting to DB to listen on sdk_tip NOTIFY');
+
+          // In case of error opening the DB client, just retry after 1".
+          return setTimeout(addListener, 1000).unref();
+        }
+
+        logger.info('Connected to DB to listen on sdk_tip NOTIFY');
+
+        // Set the function to close the client; used here for error handling and later by server close
+        this.closeNotify = () => {
+          logger.info('Closing DB connection listening on sdk_tip NOTIFY');
+          // Resets itself, in case it is called more than once...
+          this.closeNotify = noop;
+          done();
+        };
+
+        // Attach the handlers to the DB client events
+
+        client.on('end', () => {
+          const { closing } = this;
+
+          // Being this the client closed event handler, there is no longer need for the function to close it
+          this.closeNotify = noop;
+          // Set the flag for the health check
+          this.notifyConnected = false;
+
+          // If the connection was closed because the server is being closed as well, that's all
+          // For any other reason the connection was closed, just retry a new connection
+          if (!closing) addListener();
+        });
+
+        client.on('error', (e) => {
+          logger.error(e, 'Async error from sdk_tip NOTIFY');
+          this.closeNotify();
+        });
+
+        // eslint-disable-next-line unicorn/consistent-destructuring
+        client.on('notification', this.createOnNotification());
+
+        // Issue the LISTEN command to get the notification event
+        client.query('LISTEN sdk_tip', (e) => {
+          // If there was no errors in the statement, set the flag for the health check
+          if (!e) return (this.notifyConnected = true);
+
+          logger.error(e, 'Error while listening on sdk_tip NOTIFY');
```

Comment:
This is a fatal error, and we might want to fail the service altogether. (not essential for the PoC)

Reply:
It is actually a retriable error. The retry strategy is to close the connection to open a new 1 second later.